flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java

```diff
@@ -133,9 +134,13 @@ public void addState(final State<T> state) {
 		final int numberComputationStates = computationStates.size();
 		final Collection<Map<String, T>> result = new ArrayList<>();
 		final Collection<Tuple2<Map<String, T>, Long>> timeoutResult = new ArrayList<>();
+		boolean skipOtherStates = false;
 
 		// iterate over all current computations
 		for (int i = 0; i < numberComputationStates; i++) {
+			if (skipOtherStates) {
+				continue;
+			}
```

Comment:
I think you cannot simply skip the current set of computation states because then you would retain those states, which have a next condition and would thus be filtered out in the `MatchingBehaviour.AFTER_LAST` case.